Trace cli in go

turkenh opened the pull request:
PR for https://github.com/crossplaneio/crossplane/issues/744

To do:

- [x] Test and adapt "Simple Resource Class Selection" https://github.com/crossplaneio/crossplane/pull/926
- [x] Better identifier, UID missing if object does not exist
- [x] Adapt bootstrap.sh to include trace
- [x] Ready/Pending/Missing states to highlight what is wrong
- [x] Unit tests
- [x] Test different scenarios and edge case handling 
- [ ] Refactor: better naming, using constants, comments etc
- [x] Polish graph output

Build & Install:
```
$ make build
$ make install
```

Trace a kubernetesapplication:
```
$ kubectl crossplane trace kubernetesapplication <name> -n <namespace>
```

Graph output:
```
# Graphviz needs to be installed first, e.g. brew install graphviz
$  kubectl crossplane trace kubernetesapplication <name> -n <namespace> -o dot |dot -Tpng > /tmp/output.png&& open /tmp/output.png
```

turkenh committed:
trace: add crossplane-cli go project

First commit adding the golang project for trace cli

Signed-off-by: Hasan Turken <turkenh@gmail.com>

suskin commented on bin/kubectl-crossplane-trace:
```diff
@@ -0,0 +1,22 @@
+set -e
+
+function usage {
+  echo "Usage: kubectl crossplane trace KIND NAME [-n| --namespace NAMESPACE] [-h|--help]" >&2 >&2
```
It seems like `>&2 >&2` should probably be `>&2`

suskin commented on bin/kubectl-crossplane-trace:
```diff
@@ -0,0 +1,22 @@
+
+check_help "${1}"
+
+crossplane-cli trace "${@}"
```
Where does `crossplane-cli` come from?

turkenh replied:
This will be the built binary of go code. 
e.g.
```
go build -o /bin/crossplane-cli cmd/main.go
```

suskin commented on bin/kubectl-crossplane-trace:
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env bash
```
What is the benefit of having a `bash` wrapper around another binary?

turkenh replied:
As you know, kubectl plugins require an exact name matching the command/subcommand.
This way, we can use a single binary produced from the go code for different subcommands (currently only for trace but later we can consider adding more).

suskin reviewed (commented):
@turkenh I like that there's an early draft for early review and comments! A couple questions.

Would you be able to use more upbound-style commit messages? Here's the formula I've been using:

```
category: one-line summary of change

This section down here is longer, and is an explanation
of why the change is being made, or why it is happening
in the way it is being done. The goal of this section is to
help people get into the mindset of the original author, so
that they can keep making changes in the way that the
original author would be, if that's what they want to do.

commit signature
```

What testing has been done?

turkenh commented:
> Would you be able to use more upbound-style commit messages? Here's the formula I've been using

Sure. I'll most probably, overwrite the commit history here with some sort of "Initial commit" for trace cli go project since the initial commits include experimental changes and I didn't follow well scoped commits.

> What testing has been done?

I am mainly testing with wordpress stack example both happy path and also by intentionally breaking some stuff.

turkenh committed:
trace: refactor - define vars for accessing fields

We need the path of the fields for accessing the data in
unstructured object. This commit defines and use some varaibles
for this purpose.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: add relation from provider to credsecret

Now the cred secret used by a provider is also included into the
trace.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: support type.group as input

When there are two kinds from different groups, it was not possible
to indicate which one is expected to be traced. e.g. if both
"providers.gcp.crossplane.io" and "providers.aws.crossplane.io",
kubectl get providers simply selects one of them. This commit adds
support for types (or kinds) with group just like kubectl can do.
example:
kubectl crossplane trace providers.gcp.crossplane.io gcp-provider -n gcp

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: refactor - return object details instead of formatted string

This commit introduces a new method (with unittests) on trace.crossplane
interface (yet to be consumed by printer) which returns ObjectDetails
struct so that printers can access and print details with their format.
Previously, details of crossplane objects was formatted strings which
means embedding formatting logic into non related objects.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: refactor - simple printer with common template

With this commit, simple printer uses ObjectDetails returned by
crossplane objects and prints details with a common go template.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: add help message

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: delete bin/kubectl-crossplane-trace bash file

Removes the wrapper bash file which was intented to make future
subcommands available from single binary. Because it is
not necessary right now and we can consider again once we introduce
new subcommands.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: adapt bootstrap.sh for trace subcommand installation

Released artifacts will be platform specific .tar.gz files including
bin directory with all existing bash scripts plus
kubectl-crossplane-trace binary built from go code. ./bootstrap.sh
will fetch platform specific artifact and extract to appropriate
directory as before. For backward compatibility, if RELEASE is set
to master, then bootstrap will behave as before but missing trace.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: define IsReady per object and show in output

Each crossplane object now has an IsReady method indicating its
readiness. This information is also made avilable on simpleprinter
with a sign (ready/notReady/missing) and on graph with colors
(red missing, orange notReady). This could help to immediately
identify what can be wrong.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: load all auth plugins

Load all auth plugins which makes the tool able to authenticate
with cloud providers.
https://github.com/kubernetes/client-go/tree/master/examples#auth-plugins

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: add build and clean makefile targets

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: add unit tests from graphbuilder

Adds unit tests for fetchObj and BuildGraph in graphbuilder.go
Also includes some related fixing and refactorings.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: unit test for build graph with one reference

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: add comments for exporteds in trace package

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh commented:
@suskin I still have something to do, but believe that PR is ready for review.

It would be great if you could have a chance to test on your side as well.

turkenh committed:
trace: compatibility with simple resource class selection

With v0.4 Portable classes being removed and resource classes,
providers and managed resources are being changed as cluster scoped.
This commitadds anticipated compatibility changes by still trying
to keep it backward compatible with v0.3

Signed-off-by: Hasan Turken <turkenh@gmail.com>

suskin commented on bootstrap.sh:
```diff
@@ -10,13 +10,23 @@ if [[ -z "${RELEASE}" ]]; then
   RELEASE=master
 fi
 
+PLATFORM="linux"
```
@turkenh what platform are you running? I'm testing this on darwin.

turkenh replied:
I'm also using darwin, but just tested on a debian instance and working fine.

suskin replied:
kk

suskin commented on Makefile:
```diff
@@ -4,8 +4,16 @@ default:
+	ln -si $(abspath bin/kubectl-crossplane-*) $(INSTALL_DIR)/
 .PHONY: install
 
 uninstall:
```
Is the `.DS_Store` file below needed? Should we remove it and .gitignore it?

turkenh replied:
No, definitely not needed. Deleted and added to .gitignore.

suskin commented on bootstrap.sh:
```diff
@@ -10,13 +10,23 @@ if [[ -z "${RELEASE}" ]]; then
+  curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-generate_install https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-generate_install >/dev/null
+  curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-list https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-list >/dev/null
+else
+  curl -sL https://github.com/crossplaneio/crossplane-cli/releases/download/"${RELEASE}"/crossplane-cli-"${RELEASE}"-"${PLATFORM}".tar.gz | tar xz --strip 1 -C "${PREFIX}"/bin
```
Has this been tested on linux and darwin?

turkenh replied:
Yes, works fine on both. Tested with my test releases: https://github.com/turkenh/crossplane-cli/releases/tag/v0.1.2-test

We need to follow the naming convention here with our releases.

`crossplane-cli-"${RELEASE}"-"${PLATFORM}".tar.gz`

suskin commented on bootstrap.sh:
```diff
@@ -10,13 +10,23 @@ if [[ -z "${RELEASE}" ]]; then
-curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-generate_install https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-generate_install >/dev/null
-curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-list https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-list >/dev/null
+if [[ "${RELEASE}" == "master" ]]; then
+  echo "trace subcommand will not be available from master, RELEASE must be set to a released version."
```
I have been taking the approach of sending progress and error messages to stderr instead; let's redirect this: `>&2`

Also, could we put some sort of attention-grabbing prefix at the beginning of the message, and add an example of a released version? For example:

```
echo "NOTICE: the trace command is not available from master. RELEASE must be set to a released version (such as v0.2.0). See https://github.com/crossplaneio/crossplane-cli/releases for the full list of releases." >&2
```

turkenh replied:
Both makes sense. Done.

suskin commented on bootstrap.sh:
```diff
@@ -10,13 +10,23 @@ if [[ -z "${RELEASE}" ]]; then
 fi
 
+PLATFORM="linux"
+if [[ "$OSTYPE" == "darwin"* ]]; then
```
Style nitpick: could we change this to use curly braces for consistency? So something more like `"${OSTYPE}"`. I like using curly braces for clarity for scripts. In general, I like to use the long forms of syntax in scripts, since they're usually clearer.

suskin commented on bootstrap.sh:
```diff
@@ -10,13 +10,23 @@ if [[ -z "${RELEASE}" ]]; then
+  curl -sL -o "${PREFIX}"/bin/kubectl-crossplane-stack-list https://raw.githubusercontent.com/crossplaneio/crossplane-cli/"${RELEASE}"/bin/kubectl-crossplane-stack-list >/dev/null
+else
+  curl -sL https://github.com/crossplaneio/crossplane-cli/releases/download/"${RELEASE}"/crossplane-cli-"${RELEASE}"-"${PLATFORM}".tar.gz | tar xz --strip 1 -C "${PREFIX}"/bin
+fi
```
Style nitpick: I would normally put a blank line after an `if` block

suskin commented on cmd/trace/main.go:
```diff
@@ -0,0 +1,105 @@
+	var namespace string
+	var outputFormat string
+	if home := homedir.HomeDir(); home != "" {
+		pflag.StringVar(&kubeconfig, "kubeconfig", filepath.Join(home, ".kube", "config"), "(optional) Absolute path to the kubeconfig file")
```
I don't know enough about CLI libraries in go enough to have an opinion here, but is this the library that kubectl uses? I would default to trying to use the same stuff as kubectl.

suskin commented on cmd/trace/main.go:
```diff
@@ -0,0 +1,105 @@
+}
+
+func printHelp() {
+	fmt.Fprintf(os.Stderr, `Usage: kubectl crossplane trace TYPE[.GROUP] NAME [-n| --namespace NAMESPACE] [-h|--help]
```
Usability: it'd probably make sense to also include `-o`  in the usage line, or something like `[options ...]`

suskin commented on cmd/trace/main.go:
```diff
@@ -0,0 +1,105 @@
+	resource := pflag.Arg(0)
+	resourceName := pflag.Arg(1)
+	if resource == "" || resourceName == "" {
+		failWithMessage("missing arguments: SUBCOMMAND TYPE[.GROUP] NAME [-n|--namespace NAMESPACE]")
```
I'm noticing this message doesn't match the one printed by usage . . . would it possible for us to reuse the same message? What do you think of the pattern of printing the full usage when the arguments are wrong? That's usually the pattern I go with

suskin commented on cmd/trace/main.go:
```diff
@@ -0,0 +1,105 @@
+	if resource == "" || resourceName == "" {
+		failWithMessage("missing arguments: SUBCOMMAND TYPE[.GROUP] NAME [-n|--namespace NAMESPACE]")
+	}
+	fmt.Fprintf(os.Stderr, "Collecting information for %s %s in namespace %s...\n\n", resource, resourceName, namespace)
```
Nice use of stderr here

suskin commented on pkg/crossplane/application.go:
```diff
@@ -0,0 +1,79 @@
+)
+
+type Application struct {
+	u *unstructured.Unstructured
```
In general I like descriptive variable names over single-letter names, but if this is the go idiom for untyped and anonymous variables like this, I'm okay with it.

turkenh replied:
Go idiom likes short names but not sure this applies here. 
Changed it as u -> instance which is already something that I used somewhere else in the code.

suskin commented on pkg/crossplane/application.go:
```diff
@@ -0,0 +1,79 @@
+	return o.GetStatus() == stateAppSubmitted
+}
+
+func (o *Application) GetRelated(filterByLabel func(metav1.GroupVersionKind, string, string) ([]unstructured.Unstructured, error)) ([]*unstructured.Unstructured, error) {
```
Something to keep in mind for the future is that it'd be great if we could add support for more types without having to recompile the `trace` binary. We don't have to have a solution now, but it's something to think about.

turkenh replied:
Yes, that's a good point.
This is something that I also considered in the context of making this tool as a "general purpose" tracing tool as you mentioned in the https://github.com/crossplaneio/crossplane/issues/744. This could be possible with some sort of configuration file which describes the types and their relations with other types.

suskin commented on pkg/crossplane/application_test.go:
```diff
@@ -0,0 +1,108 @@
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+func TestApplication_GetObjectDetails(t *testing.T) {
```
Awesome! Could we also add a `make test` target for running the unit tests?

suskin commented on pkg/crossplane/application.go:
```diff
@@ -0,0 +1,79 @@
+}
+
+func (o *Application) GetStatus() string {
+	return getNestedString(o.u.Object, fieldsStatusState...)
```
I usually prefer that the imports for shared functionality are explicit so they're easier to follow for readers who are not familiar with the codebase, but this is something we can probably reexamine in the future.

turkenh replied:
Well, if you mean `getNestedString` this is not coming from an import. it is a helper in the same package.
Otherwise, I agree with your comment in general.

suskin commented on pkg/trace/graphprinter.go:
```diff
@@ -0,0 +1,60 @@
+}
+
+// Print prints graph definition in Graphviz dot format for a given root node.
+func (p *GraphPrinter) Print(nodes []*Node) error {
```
Yesss

suskin reviewed (approved):
Awesome - thanks @turkenh ! Since it's a larger pull request it's a little harder to do an in-depth review quickly, but it looks pretty good to me. I just had some minor questions and suggestions.

I'm going to approve this since we're in different timezones, but I'd like us to at a minimum address the comments about testing (add `make test`, verify that installation from a release has been tested) and usability before we merge the changes.

For posterity, I asked @turkenh in a separate conversation whether we would be able to use more of the code from kubectl - for example, for the `kubectl get` commands. He said that he looked, but that he wasn't able to find anything which was easily reusable. We both agreed that we'd like to revisit that in the future at some point, but that we're okay with doing a quick thing for now so we can start using the tool.

suskin commented:
I forgot to mention, but I did some integration testing on my side as well. I used it to inspect the statuses for a mysqlinstance and a kubernetesapplication which I created as part of the Stacks guide. And I used it to watch the kubernetesapplication become ready over time. It was way better than doing those things without the tool!

As another note, we'll want to update some of the documentation - in [this repo](https://github.com/crossplaneio/crossplane-cli/blob/master/README.md) and in [crossplane](https://github.com/crossplaneio/crossplane/blame/518648b051cd2055fb2d95dbdbc120855af7eedf/docs/stacks-guide.md#L202) - to recommend that we use `trace`. The documentation will need to be updated before we consider the `trace` work to be complete.

turkenh committed:
add .gitignore

Adds .DS_Store (a mac os system file) and produced binary to
.gitignore.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
bootsrap: style and message improvements

Attention grabbing message for installations from master and
curly braces for variables.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: rename unstructured variables to more descriptive

Changed as "instance" from "u"

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
makefile: add test target for go unit tests

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh commented:
@suskin Thanks a lot for reviewing and testing!
All of your comments should be resolved right now.

> For posterity, I asked @turkenh in a separate conversation whether we would be able to use more of the code from kubectl - for example, for the kubectl get commands. He said that he looked, but that he wasn't able to find anything which was easily reusable. We both agreed that we'd like to revisit that in the future at some point, but that we're okay with doing a quick thing for now so we can start using the tool.

Most related parts from `kubectl get` was the ones for printing, since for interacting with api-server I used client-go dynamic client which was pretty good. I checked https://github.com/kubernetes/cli-runtime which could be helpful, but as @suskin mentioned, I found adding my own simple printer as more easy and flexible. I still agree that this should be something to revisit in future.

> I forgot to mention, but I did some integration testing on my side as well. I used it to inspect the statuses for a mysqlinstance and a kubernetesapplication which I created as part of the Stacks guide. And I used it to watch the kubernetesapplication become ready over time. It was way better than doing those things without the tool!

Thanks a lot!
Which cloud provider were you using? I always used GCP and now looking for a way to test with others. Another this that needs to be test is the recent changes happening around simple resource class selection.

turkenh committed:
trace: fixes for simple class selection

Adapt crossplane object to work with both v0.3 and v0.4

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: better error message in case object not found

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: ux improvement - put overview to bottom

Previously overview was printed first and details second. But for
long outputs, user first sees the bottom of details as console
slided down. With this commit, overview will be what user sees
first.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: respect KUBECONFIG env var

With this change kubeconfig is selected with the same logic as
kubectl.
--kubeconfig flag overwrites all
env KUBECONFIG selected otherwise
default path (~/.kube/config) in homedir is none set

Signed-off-by: Hasan Turken <turkenh@gmail.com>

suskin commented:
> Which cloud provider were you using?

I was using GCP, and I was also using Crossplane 0.3 instead of master

suskin commented on pkg/trace/graphbuilder.go:
```diff
@@ -0,0 +1,187 @@
+	err = g.fetchObj(root)
+	if kerrors.IsNotFound(err) {
+		return root, nil, errors.New(
+			fmt.Sprintf("Object to trace is not found: \"%s\" \"%s\" in namespace \"%s\"", groupRes, name, namespace))
```
Grammar nitpick: `Object to trace not found` or `Object to trace was not found` would be better. I would normally go with the `Object to trace not found` form

suskin commented on pkg/trace/graphbuilder_test.go:
```diff
@@ -0,0 +1,365 @@
+				resource:          "kubernetescluster",
+			},
+			want: want{
+				err: errors.NewNotFound(schema.ParseGroupResource("kubernetesclusters.compute.crossplane.io"), "test"),
```
When I ran the unit tests, I got this:

```
--- FAIL: TestKubeGraphBuilder_BuildGraph (0.00s)
    --- FAIL: TestKubeGraphBuilder_BuildGraph/NotExist (0.00s)
        graphbuilder_test.go:306: g.BuildGraph(...): -want error, +got error:   interface{}(
            - 	e`kubernetesclusters.compute.crossplane.io "test" not found`,
            + 	e`Object to trace is not found: "kubernetescluster" "test" in namespace "testnamespace"`,
              )
```

turkenh replied:
Fixed. I should get used to running make test before pushing :)

suskin replied:
haha : ) - ideally we'll also set up some automated builds so we don't need to always remember

turkenh committed:
trace: print additional status details one per line

Previously it was shown as tabbed but on AWS and Azure
resources value does not fit into colums which breaks
layout

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: change wording in error message and fix unit test

From "Object to trace is not found" to "Object to trace not found"
as per review comments.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

turkenh committed:
trace: fix finding related provider of a resource class

Full object reference from resource class to provider is not available.
Only name of the provider there. To resolve the provider, there was a
workaround in the code, which sets apiVersion of the provider as the
apiVersion of the resource class. This assumption does not hold with
v1beta1 versions recently introduced in stack-gcp. This commit fixes
by setting group instead of apiVersion.

Signed-off-by: Hasan Turken <turkenh@gmail.com>

suskin reviewed (approved):
Thanks @turkenh ! This looks good to me. I believe we'll need to do a couple more things to get a release up, but we can work on it after this pull request is merged.
